example/quadratic.f90

```diff
@@ -0,0 +1,170 @@
+  ypred_batch_sgd = [(net_batch_sgd % predict([xtest(i)]), i = 1, test_size)]
+  ypred_minibatch_sgd = [(net_minibatch_sgd % predict([xtest(i)]), i = 1, test_size)]
+
+  ! Print the mean squared error
+  print '("Stochastic gradient descent MSE:", f9.6)', sum((ypred_sgd - ytest)**2) / size(ytest)
+  print '("     Batch gradient descent MSE: ", f9.6)', sum((ypred_batch_sgd - ytest)**2) / size(ytest)
+  print '(" Minibatch gradient descent MSE: ", f9.6)', sum((ypred_minibatch_sgd - ytest)**2) / size(ytest)
+
+contains
+
+   subroutine sgd_optimizer(net, x, y, learning_rate, num_epochs)
+    ! In the stochastic gradient descent (SGD) optimizer, we run the forward
+    ! and backward passes and update the weights for each training sample,
+    ! one at a time.
+    type(network), intent(inout) :: net
+    real, dimension(:), intent(in) :: x, y
+    real, intent(in) :: learning_rate
+    integer, intent(in) :: num_epochs
+    integer :: i, n
+    integer, dimension(:), allocatable :: indices
+
+    print *, "Running SGD optimizer..."
+
+    ! Generate shuffled indices for the mini-batches
+    allocate(indices(size(x)))
+    indices = [(i, i = 1, size(x))]
+    call shuffle(indices)
+
+    do n = 1, num_epochs
+      do i = 1, size(x)
+        call net % forward([x(indices(i))])
+        call net % backward([y(indices(i))])
+        call net % update(learning_rate)
+      end do
+    end do
+
+  end subroutine sgd_optimizer
+
+
+  subroutine batch_gd_optimizer(net, x, y, learning_rate, num_epochs)
+    ! Like the stochastic gradient descent (SGD) optimizer, except that here we
+    ! accumulate the weight gradients for all training samples and update the
+    ! weights once per epoch.
+    type(network), intent(inout) :: net
+    real, dimension(:), intent(in) :: x, y
+    real, intent(in) :: learning_rate
+    integer, intent(in) :: num_epochs
+    integer :: i, n
+
+    print *, "Running batch GD optimizer..."
+
+    do n = 1, num_epochs
+      do i = 1, size(x)
+        call net % forward([x(i)])
+        call net % backward([y(i)])
+      end do
+      call net % update(learning_rate / size(x))
+    end do
+
+  end subroutine batch_gd_optimizer
+
+
+  subroutine minibatch_gd_optimizer(net, x, y, learning_rate, num_epochs, batch_size)
+    ! Like the batch SGD optimizer, except that here we accumulate the weight
+    ! over a number of mini batches and update the weights once per mini batch.
+    type(network), intent(inout) :: net
+    real, dimension(:), intent(in) :: x, y
+    real, intent(in) :: learning_rate
+    integer, intent(in) :: num_epochs, batch_size
+    integer :: i, j, n, num_samples, num_batches, start_index, end_index
+    real, dimension(:), allocatable :: batch_x, batch_y
+
+    print *, "Running mini-batch GD optimizer..."
+
+    num_samples = size(x)
+    num_batches = num_samples / batch_size
+
+    allocate(batch_x(batch_size), batch_y(batch_size))
+
+    do n = 1, num_epochs
```

Comment:
@Spnetic-5 sorry I mistyped the SGD subroutine in my earlier comment. I meant that the mini-batches should be shuffled here in the minibatch GD :).

Reply:
Yeah, even I got confused as there are no batches in SGD, I'll update the changes

Reply:
It's the same approach as what you already did, but just shuffling the start_indices for mini-batches rather than all indices in the dataset.